wp_api/src/middleware.rs

```diff
@@ -0,0 +1,364 @@
+    async fn process(
+        &self,
+        request_executor: Arc<(dyn RequestExecutor + 'static)>,
+        response: WpNetworkResponse,
+        request: Arc<WpNetworkRequest>,
+    ) -> Result<WpNetworkResponse, WpApiError> {
+        let mut retry_count = 0;
+        let mut response = response;
+
+        loop {
+            if retry_count >= self.max_retries {
+                return Err(WpApiError::RequestExecutionFailed {
+                    status_code: Some(response.status_code),
+                    redirects: None,
+                    reason: RequestExecutionErrorReason::MisconfiguredRateLimitError {},
+                });
+            }
+
+            if response.is_rate_limit_exceeded() && self.max_retries != 0 {
+                let retry_after = response.get_retry_after();
+
+                if let Some(mut retry_after) = retry_after {
+                    // If the server sends some super-long value, we don't want to wait that long
+                    if retry_after > self.max_retry_wait_seconds {
+                        retry_after = self.max_retry_wait_seconds;
+                        async_sleep(Duration::from_secs(retry_after)).await;
+                    }
+
+                    // task::sleep(Duration::from_secs(retry_after)).await;
+                    response = request_executor.execute(request.clone()).await?;
+                } else {
+                    return Ok(response); // It's not ok, but we'll let the layer above handle that – we have no idea how long to wait so we shouldn't try
+                }
+                retry_count += 1;
+            }
+        }
+    }
+}
+
+fn async_sleep(duration: std::time::Duration) -> impl Future<Output = ()> {
+    let (tx, rx) = futures::channel::oneshot::channel::<i32>();
+
+    thread::spawn(move || {
+        thread::sleep(duration);
+        let _ = tx.send(0); // ignore error because it is perfectly ok when the receiver is dropped
+    });
+
+    async move {
+        rx.await.unwrap();
+    }
+}
+
+// MARK: - HttpAuthenticationMiddleware
+
+#[derive(Debug, uniffi::Object)]
+struct HttpAuthenticationMiddleware {
+    username: String,
+    password: String,
+}
+
+#[uniffi::export]
+impl HttpAuthenticationMiddleware {
+    #[uniffi::constructor]
+    fn new(username: String, password: String) -> Self {
+        println!("Creating HTTP authentication middleware");
+        Self { username, password }
+    }
+}
+
+#[uniffi::export]
+#[async_trait::async_trait]
+impl WpApiMiddleware for HttpAuthenticationMiddleware {
+    async fn process(
+        &self,
+        request_executor: Arc<(dyn RequestExecutor + 'static)>,
+        response: WpNetworkResponse,
+        request: Arc<WpNetworkRequest>,
+    ) -> Result<WpNetworkResponse, WpApiError> {
+        if !response.is_http_authentication_required() {
+            return Ok(response);
+        }
+
+        let new_request = request.adding_http_authentication(&self.username, &self.password);
```

Comment:
When using `HttpAuthenticationMiddleware` (instead of adding the header when building requests), we would send two requests for endpoints that require authentication: one without the authentication header and the other with the authentication header.

What do you think about allowing handling outgoing requests, too, so that we can insert headers before sending requests?

Reply:
Also, it's best to check request.url first before adding the header. We don't want to send the application password to any random sites.